pkg/adaptation/adaptation.go

```diff
@@ -334,12 +334,27 @@ func (r *Adaptation) startPlugins() (retErr error) {
 		}
 
 		if err := p.start(r.name, r.version); err != nil {
-			return err
+			return fmt.Errorf("failed to start NRI Plugin %q: %w", name, err)
 		}
 
 		plugins = append(plugins, p)
 	}
 
+	syncPlugins := func(ctx context.Context, pods []*PodSandbox, containers []*Container) ([]*ContainerUpdate, error) {
+		var updates []*ContainerUpdate
+		for _, plugin := range plugins {
+			us, err := plugin.synchronize(ctx, pods, containers)
+			if err != nil {
+				return nil, fmt.Errorf("failed to sync NRI Plugin %q: %w", plugin.name(), err)
```

Comment:
should this be a warning instead and continue?

Reply:
There are two possible approaches to consider: skipping the synchronization if it fails or returning an error.

I chose to return an error for the following reason: `pre-install plugins` and `external plugins` are different in some ways. 
The `pre-install plugin binary` is launched by the NRI adaptation, so it is difficult for users to know whether the binary plugin has been launched successfully, apart from the logs. 
Therefore, users may overlook the fact that the `plugin binary` is not actually working. That's why I require that all `pre-install plugins` requested by the user must be launched(started, synced) successfully. 
This is similar to starting and returning an error if it fails.
https://github.com/containerd/nri/blob/fa3ab8fbf3c5e58f652a7a54f750f9b6a02416f7/pkg/adaptation/adaptation.go#L336-L338

On the other hand, external plugins are usually registered by external binaries, and we could perceive more clearly whether the NRI registration and synchronization are successful or not.

Reply:
I suppose two other patterns would be:
3)  to introduce config/management for requiring/ordering (add requires logic), auto restarting etc.. for plugin start && synch
4) attempt to start and synch all and report all errors

that said should we stop any that we started if there's an error..

Reply:
> 3) to introduce config/management for requiring/ordering (add requires logic), auto restarting etc.. for plugin start && synch

Could you introduce more details? thanks

Reply:
current containerd plugins have a requires section for specifying which must plugins be init'd.. before initing a plugin

we probably need an additional state beyond init for the container plugin that would cover ready for work.. for example in the cri case ready may mean all containers/pods restarted ready for grpc calls

nri might need a requires list for the plugins (unless we go with external management via systemd? or whatnot) and then within that you could have the concept of ready which might at least cover init and sync ^  

additionally we might want to have a flag in the plugin config regarding if sync is required for "ready"

Reply:
@dmcgowan thoughts?